File: src/sentinel/detectors/instruction_override.py
"""Detects attempts to countermand earlier instructions.

The discriminating idea is that an override attempt pairs an *imperative verb*
with a *reference to prior context* ("ignore ... previous instructions"). Simply
matching the verb produces false positives on ordinary prose such as
"ignore the noise in the dataset", so the verb alone is never enough.

Two deliberate choices keep the false-positive rate down:

* Only base-form verbs are matched. "How do I stop users from overriding my
  system prompt?" is a legitimate developer question; matching ``\\boverride\\b``
  will not fire on the gerund "overriding".
* The reference to prior context must appear within a short window after the
  verb, so an unrelated "previous" elsewhere in the sentence does not trigger.
"""

import re

from sentinel.types import Signal

# Base forms only. Gerunds and participles are excluded on purpose; see module docstring.
_VERB = r"(?:ignore|disregard|forget|override|bypass|discard|skip|omit)"

# A reference to prior context, or to the instruction set itself.
_TRIGGER = (
    r"(?:previous|prior|above|earlier|preceding|initial|original|foregoing"
    r"|everything|instructions?|directives?|rules?|guidelines?|constraints?"
    r"|restrictions?|policies|policy)"
)

# The object of the override, when the trigger was only a modifier
# ("previous" in "ignore previous instructions").
_OBJECT = r"(?:instructions?|directives?|rules?|guidelines?|prompts?|context|messages?)"

# Verb, then a reference to prior context within three words, optionally
# followed by the thing being overridden so the reported evidence reads as a
# whole phrase rather than a fragment. The bounded repetition keeps this
# ReDoS-safe.
_OVERRIDE_RE = re.compile(
    rf"\b{_VERB}\b(?:\s+\w+){{0,3}}?\s+\b{_TRIGGER}\b(?:\s+{_OBJECT}\b)?",
    re.IGNORECASE,
)

# "new instructions:" style redirection, which carries no verb.
_REDIRECT_RE = re.compile(
    r"\bnew\s+(?:instructions?|directives?|rules?|task|prompt)\s*[:\-]",
    re.IGNORECASE,
)


class InstructionOverride:
    """Flags text that tries to countermand prior instructions."""

    name = "InstructionOverride"

    def scan(self, text: str) -> list[Signal]:
        if not text:
            return []

        signals: list[Signal] = []
        for pattern, weight in ((_OVERRIDE_RE, 0.85), (_REDIRECT_RE, 0.7)):
            for match in pattern.finditer(text):
                signals.append(
                    Signal(
                        detector=self.name,
                        weight=weight,
                        evidence=match.group(0),
                        span=match.span(),
                    )
                )
        return signals

File: web/src/lib/sentinel/detectors/instructionOverride.ts
/**
 * Detects attempts to countermand earlier instructions.
 * Mirrors `src/sentinel/detectors/instruction_override.py`.
 *
 * ─── REFERENCE IMPLEMENTATION ────────────────────────────────────────────
 * This is the worked example. The other four detectors follow the same shape:
 *
 *   1. Module-level regexes, compiled once (never build a regex inside scan).
 *   2. A class with a `name` matching its DetectorName, and `scan()`.
 *   3. Return `[]` for empty input, early.
 *   4. One Signal per match, carrying evidence and span for UI highlighting.
 *   5. Bounded quantifiers only, so the pattern cannot backtrack catastrophically.
 * ─────────────────────────────────────────────────────────────────────────
 *
 * The discriminating idea is that an override attempt pairs an *imperative
 * verb* with a *reference to prior context* ("ignore ... previous
 * instructions"). Matching the verb alone produces false positives on ordinary
 * prose such as "ignore the noise in the dataset".
 *
 * Only base-form verbs are matched. "How do I stop users from overriding my
 * system prompt?" is a legitimate developer question, and `\boverride\b` does
 * not fire on the gerund "overriding".
 */

import type { Detector, DetectorName, Signal } from "../types";

const VERB = "(?:ignore|disregard|forget|override|bypass|discard|skip|omit)";

const TRIGGER =
  "(?:previous|prior|above|earlier|preceding|initial|original|foregoing" +
  "|everything|instructions?|directives?|rules?|guidelines?|constraints?" +
  "|restrictions?|policies|policy)";

/** The object of the override, so reported evidence reads as a whole phrase
 *  ("ignore previous instructions") rather than a fragment ("ignore previous"). */
const OBJECT =
  "(?:instructions?|directives?|rules?|guidelines?|prompts?|context|messages?)";

const OVERRIDE_RE = new RegExp(
  `\\b${VERB}\\b(?:\\s+\\w+){0,3}?\\s+\\b${TRIGGER}\\b(?:\\s+${OBJECT}\\b)?`,
  "gi",
);

/** "new instructions:" style redirection, which carries no verb. */
const REDIRECT_RE = /\bnew\s+(?:instructions?|directives?|rules?|task|prompt)\s*[:\-]/gi;

export class InstructionOverride implements Detector {
  readonly name: DetectorName = "InstructionOverride";

  scan(text: string): Signal[] {
    if (!text) return [];

    const signals: Signal[] = [];
    const patterns: ReadonlyArray<readonly [RegExp, number]> = [
      [OVERRIDE_RE, 0.85],
      [REDIRECT_RE, 0.7],
    ];

    for (const [pattern, weight] of patterns) {
      // Reset lastIndex: these are module-level /g regexes reused across calls,
      // and a stale lastIndex would silently skip matches on the next scan.
      pattern.lastIndex = 0;
      for (const match of text.matchAll(pattern)) {
        signals.push({
          detector: this.name,
          weight,
          evidence: match[0],
          span: [match.index, match.index + match[0].length],
        });
      }
    }
    return signals;
  }
}
